File: python-bridge/ast_to_sexp.py
#!/usr/bin/env python3
"""
Frankenstein Python bridge — AST → S-expression dumper.

Reads a Python source file from argv[1] (or stdin if "-") and prints a
single-line S-expression representing a small subset of the AST that the
Haskell-side Python bridge knows how to translate to OrganIR.

Supported subset (enough for arithmetic / recursive integer programs
plus simple records):

  Module(stmts)
  FunctionDef(name, params, body)
  ClassDef(name, fields)           -- record-like: __init__ with self.X = X
  Return(expr|None)
  If(test, then-stmts, else-stmts)
  Assign(target_name, expr)        -- single Name target only
  ExprStmt(expr)
  BinOp(op_name, lhs, rhs)         -- Add Sub Mult FloorDiv Mod
  UnaryOp(op_name, operand)        -- USub
  Compare(lhs, op_name, rhs)       -- Eq NotEq Lt LtE Gt GtE  (single op)
  Constant(int|str|bool)
  Name(id)
  Attr(value, attr)                -- value.attr
  Call(callee_name, args)          -- callee must be a bare Name
  Raise(expr)                      -- raise expr (expr is required)
  Try(body-stmts, except-stmts)    -- single bare `except:` only;
                                   --   no type, no `as`, no else/finally

Anything else raises NotImplementedError.
"""

from __future__ import annotations

import ast
import sys


def quote(s: str) -> str:
    out = ['"']
    for ch in s:
        if ch == '\\':
            out.append('\\\\')
        elif ch == '"':
            out.append('\\"')
        elif ch == '\n':
            out.append('\\n')
        elif ch == '\t':
            out.append('\\t')
        else:
            out.append(ch)
    out.append('"')
    return ''.join(out)


def emit(node: ast.AST) -> str:
    if isinstance(node, ast.Module):
        return "(Module " + " ".join(emit(s) for s in node.body) + ")"

    if isinstance(node, ast.FunctionDef):
        params = "(" + " ".join(quote(a.arg) for a in node.args.args) + ")"
        body = "(" + " ".join(emit(s) for s in node.body) + ")"
        return f"(FunctionDef {quote(node.name)} {params} {body})"

    if isinstance(node, ast.ClassDef):
        # Record-like class: a single __init__ that does `self.X = X` for
        # each parameter. We extract the field names in declaration order.
        if node.bases or node.keywords or node.decorator_list:
            raise NotImplementedError(
                f"class {node.name}: bases/keywords/decorators not supported")
        init = None
        for stmt in node.body:
            if isinstance(stmt, ast.FunctionDef) and stmt.name == "__init__":
                init = stmt
                break
        if init is None:
            raise NotImplementedError(
                f"class {node.name}: must define __init__")
        # Skip 'self' (first arg). Field order = positional arg order.
        fields = [a.arg for a in init.args.args[1:]]
        # Sanity check the body shape: every stmt must be `self.X = X`
        # for some X in fields, in declaration order.
        seen = []
        for stmt in init.body:
            if not isinstance(stmt, ast.Assign): continue
            if len(stmt.targets) != 1: continue
            tgt = stmt.targets[0]
            if (isinstance(tgt, ast.Attribute)
                and isinstance(tgt.value, ast.Name)
                and tgt.value.id == "self"
                and isinstance(stmt.value, ast.Name)
                and stmt.value.id == tgt.attr):
                seen.append(tgt.attr)
        if seen != fields:
            raise NotImplementedError(
                f"class {node.name}: __init__ body must be only "
                f"self.X = X assignments matching the parameters")
        flds = "(" + " ".join(quote(f) for f in fields) + ")"
        return f"(ClassDef {quote(node.name)} {flds})"

    if isinstance(node, ast.Return):
        if node.value is None:
            return "(Return (None))"
        return f"(Return {emit(node.value)})"

    if isinstance(node, ast.If):
        thn = "(" + " ".join(emit(s) for s in node.body) + ")"
        els = "(" + " ".join(emit(s) for s in node.orelse) + ")"
        return f"(If {emit(node.test)} {thn} {els})"

    if isinstance(node, ast.Assign):
        if len(node.targets) != 1 or not isinstance(node.targets[0], ast.Name):
            raise NotImplementedError("only single-Name assignment targets supported")
        return f"(Assign {quote(node.targets[0].id)} {emit(node.value)})"

    if isinstance(node, ast.Expr):
        return f"(ExprStmt {emit(node.value)})"

    if isinstance(node, ast.BinOp):
        return f"(BinOp {quote(type(node.op).__name__)} {emit(node.left)} {emit(node.right)})"

    if isinstance(node, ast.UnaryOp):
        return f"(UnaryOp {quote(type(node.op).__name__)} {emit(node.operand)})"

    if isinstance(node, ast.Compare):
        if len(node.ops) != 1 or len(node.comparators) != 1:
            raise NotImplementedError("only single-operator comparisons supported")
        return (f"(Compare {emit(node.left)} "
                f"{quote(type(node.ops[0]).__name__)} {emit(node.comparators[0])})")

    if isinstance(node, ast.Constant):
        if isinstance(node.value, bool):
            # bool is subclass of int — handle first
            return f"(ConstantBool {'1' if node.value else '0'})"
        if isinstance(node.value, int):
            return f"(Constant {node.value})"
        if isinstance(node.value, str):
            return f"(ConstantStr {quote(node.value)})"
        raise NotImplementedError(f"unsupported constant: {node.value!r}")

    if isinstance(node, ast.Name):
        return f"(Name {quote(node.id)})"

    if isinstance(node, ast.Attribute):
        return f"(Attr {emit(node.value)} {quote(node.attr)})"

    if isinstance(node, ast.Raise):
        if node.exc is None:
            raise NotImplementedError("bare `raise` (no expression) not supported")
        if node.cause is not None:
            raise NotImplementedError("`raise X from Y` not supported")
        return f"(Raise {emit(node.exc)})"

    if isinstance(node, ast.Try):
        # MVP: exactly one bare `except:` clause, no `as`, no else/finally.
        if node.orelse:
            raise NotImplementedError("try/except `else:` clause not supported")
        if node.finalbody:
            raise NotImplementedError("try/except `finally:` clause not supported")
        if len(node.handlers) != 1:
            raise NotImplementedError("only a single `except:` handler supported")
        h = node.handlers[0]
        if h.type is not None:
            raise NotImplementedError(
                "typed `except SomeError:` not supported — use bare `except:`")
        if h.name is not None:
            raise NotImplementedError("`except ... as e:` binding not supported")
        body = "(" + " ".join(emit(s) for s in node.body) + ")"
        handler = "(" + " ".join(emit(s) for s in h.body) + ")"
        return f"(Try {body} {handler})"

    if isinstance(node, ast.Call):
        if not isinstance(node.func, ast.Name):
            raise NotImplementedError("only direct (Name) call targets supported")
        if node.keywords:
            raise NotImplementedError("keyword arguments not supported")
        args = "(" + " ".join(emit(a) for a in node.args) + ")"
        return f"(Call {quote(node.func.id)} {args})"

    raise NotImplementedError(f"unsupported AST node: {type(node).__name__}")


def main() -> int:
    if len(sys.argv) != 2:
        print("usage: ast_to_sexp.py <file.py|->", file=sys.stderr)
        return 2
    path = sys.argv[1]
    src = sys.stdin.read() if path == "-" else open(path, "r").read()
    try:
        tree = ast.parse(src, filename=path)
        print(emit(tree))
    except SyntaxError as e:
        print(f"SyntaxError: {e}", file=sys.stderr)
        return 1
    except NotImplementedError as e:
        print(f"NotImplementedError: {e}", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: go-bridge/ast_to_sexp.go
// Frankenstein Go bridge — AST → S-expression dumper.
//
// Usage: go run go-bridge/ast_to_sexp.go file.go
//
// Reads a Go source file, parses it via the standard library go/parser,
// and prints a single-line S-expression representing a small subset of
// the AST that the Haskell-side Go bridge knows how to translate to
// OrganIR.
//
// Supported subset (enough for arithmetic / recursive integer programs):
//
//   File          (Module name [decls])
//   FuncDecl      (FuncDecl name [params] body-block)
//   BlockStmt     (Block [stmts])
//   ReturnStmt    (Return expr | (Return (None)))
//   IfStmt        (If test then-block else-block)            -- else-block may be (Block ())
//   AssignStmt    (Assign name expr)                          -- := and = with single Ident lhs
//   ExprStmt      (ExprStmt expr)
//   BasicLit      (Constant integer)
//   Ident         (Name id)
//   BinaryExpr    (BinOp op-string lhs rhs)
//   UnaryExpr     (UnaryOp op-string operand)
//   CallExpr      (Call callee-name [args])                   -- callee must be a bare Ident
//   ParenExpr     transparent (passes through)
//
// Anything else causes an error.

package main

import (
	"fmt"
	"go/ast"
	"go/parser"
	"go/token"
	"os"
	"strings"
)

func quote(s string) string {
	var b strings.Builder
	b.WriteByte('"')
	for _, ch := range s {
		switch ch {
		case '\\':
			b.WriteString(`\\`)
		case '"':
			b.WriteString(`\"`)
		case '\n':
			b.WriteString(`\n`)
		case '\t':
			b.WriteString(`\t`)
		default:
			b.WriteRune(ch)
		}
	}
	b.WriteByte('"')
	return b.String()
}

func emit(node ast.Node) (string, error) {
	switch n := node.(type) {

	case *ast.File:
		var parts []string
		parts = append(parts, "Module", quote(n.Name.Name))
		var decls []string
		for _, d := range n.Decls {
			s, err := emit(d)
			if err != nil {
				return "", err
			}
			decls = append(decls, s)
		}
		parts = append(parts, "("+strings.Join(decls, " ")+")")
		return "(" + strings.Join(parts, " ") + ")", nil

	case *ast.FuncDecl:
		if n.Recv != nil {
			return "", fmt.Errorf("methods not supported")
		}
		var params []string
		if n.Type.Params != nil {
			for _, field := range n.Type.Params.List {
				for _, name := range field.Names {
					params = append(params, quote(name.Name))
				}
			}
		}
		body, err := emit(n.Body)
		if err != nil {
			return "", err
		}
		return fmt.Sprintf("(FuncDecl %s (%s) %s)",
			quote(n.Name.Name), strings.Join(params, " "), body), nil

	case *ast.BlockStmt:
		var stmts []string
		for _, s := range n.List {
			es, err := emit(s)
			if err != nil {
				return "", err
			}
			stmts = append(stmts, es)
		}
		return "(Block (" + strings.Join(stmts, " ") + "))", nil

	case *ast.ReturnStmt:
		if len(n.Results) == 0 {
			return "(Return (None))", nil
		}
		if len(n.Results) > 1 {
			return "", fmt.Errorf("multi-value return not supported")
		}
		e, err := emit(n.Results[0])
		if err != nil {
			return "", err
		}
		return "(Return " + e + ")", nil

	case *ast.IfStmt:
		if n.Init != nil {
			return "", fmt.Errorf("if with init clause not supported")
		}
		test, err := emit(n.Cond)
		if err != nil {
			return "", err
		}
		thn, err := emit(n.Body)
		if err != nil {
			return "", err
		}
		var els string
		if n.Else == nil {
			els = "(Block ())"
		} else {
			// Else may be a *BlockStmt or another *IfStmt. Wrap an *IfStmt
			// in a synthetic block so the Haskell side always sees a block.
			if blk, ok := n.Else.(*ast.BlockStmt); ok {
				els, err = emit(blk)
			} else {
				inner, e := emit(n.Else)
				if e != nil {
					return "", e
				}
				els = "(Block (" + inner + "))"
			}
			if err != nil {
				return "", err
			}
		}
		return fmt.Sprintf("(If %s %s %s)", test, thn, els), nil

	case *ast.AssignStmt:
		if len(n.Lhs) != 1 || len(n.Rhs) != 1 {
			return "", fmt.Errorf("only single-target assignment supported")
		}
		ident, ok := n.Lhs[0].(*ast.Ident)
		if !ok {
			return "", fmt.Errorf("assignment LHS must be an identifier")
		}
		rhs, err := emit(n.Rhs[0])
		if err != nil {
			return "", err
		}
		return fmt.Sprintf("(Assign %s %s)", quote(ident.Name), rhs), nil

	case *ast.ExprStmt:
		e, err := emit(n.X)
		if err != nil {
			return "", err
		}
		return "(ExprStmt " + e + ")", nil

	case *ast.BasicLit:
		if n.Kind != token.INT {
			return "", fmt.Errorf("only integer literals supported, got %s", n.Kind)
		}
		return "(Constant " + n.Value + ")", nil

	case *ast.Ident:
		// Built-in true/false → 1/0.
		if n.Name == "true" {
			return "(Constant 1)", nil
		}
		if n.Name == "false" {
			return "(Constant 0)", nil
		}
		return "(Name " + quote(n.Name) + ")", nil

	case *ast.BinaryExpr:
		l, err := emit(n.X)
		if err != nil {
			return "", err
		}
		r, err := emit(n.Y)
		if err != nil {
			return "", err
		}
		return fmt.Sprintf("(BinOp %s %s %s)", quote(n.Op.String()), l, r), nil

	case *ast.UnaryExpr:
		x, err := emit(n.X)
		if err != nil {
			return "", err
		}
		return fmt.Sprintf("(UnaryOp %s %s)", quote(n.Op.String()), x), nil

	case *ast.CallExpr:
		ident, ok := n.Fun.(*ast.Ident)
		if !ok {
			return "", fmt.Errorf("only direct (Ident) call targets supported")
		}
		var args []string
		for _, a := range n.Args {
			s, err := emit(a)
			if err != nil {
				return "", err
			}
			args = append(args, s)
		}
		return fmt.Sprintf("(Call %s (%s))", quote(ident.Name), strings.Join(args, " ")), nil

	case *ast.ParenExpr:
		return emit(n.X)

	case *ast.GenDecl:
		// Skip imports / non-function decls silently — return empty marker.
		return "(Skip)", nil

	default:
		return "", fmt.Errorf("unsupported AST node: %T", n)
	}
}

func main() {
	if len(os.Args) != 2 {
		fmt.Fprintln(os.Stderr, "usage: ast_to_sexp <file.go|->")
		os.Exit(2)
	}
	path := os.Args[1]

	fset := token.NewFileSet()
	var src interface{}
	if path == "-" {
		buf := make([]byte, 0, 4096)
		tmp := make([]byte, 4096)
		for {
			n, err := os.Stdin.Read(tmp)
			if n > 0 {
				buf = append(buf, tmp[:n]...)
			}
			if err != nil {
				break
			}
		}
		src = buf
		path = "<stdin>"
	}

	file, err := parser.ParseFile(fset, path, src, parser.SkipObjectResolution)
	if err != nil {
		fmt.Fprintln(os.Stderr, "ParseError:", err)
		os.Exit(1)
	}

	out, err := emit(file)
	if err != nil {
		fmt.Fprintln(os.Stderr, "EmitError:", err)
		os.Exit(1)
	}
	fmt.Println(out)
}
